composer/datasets/streaming/dataset.py

```diff
@@ -237,91 +248,145 @@ def _insert_shard_samples(self, shard: int, part_min_id: int, part_max_id: int)
+        """Download and load the given missing shards.
+
+        Args:
+            missing_shards (List[int]): The missing shards.
+            shards_to_download (List[int]): List of shards to download by this worker.
+            min_id (int): Lowest sample ID of this partition.
+            max_id (int): Highest sample ID of this partition.
+            num_processes (Optional[int]): Number of concurrent shard downloads (ie, size of process pool). If None,
+                uses number of CPUs.
+        """
+        pool = Pool(num_processes)
+        download_shard = lambda shard: self._download_shard(shard, shards_to_download)
+        try:
+            for shard in pool.imap_unordered(download_shard, missing_shards):
+                self._load_shards([shard], min_id, max_id)
+        except Exception as e:
+            with self._lock:
                 self._download_status = _DownloadStatus.FAILED
                 self._download_exception = e
-            self._insert_shard_samples(shard, part_min_id, part_max_id)
-
+            return
         with self._lock:
             self._download_status = _DownloadStatus.DONE
 
-    def __len__(self) -> int:
-        """Get the length of the dataset.
+    def _download_shards_sequential(self, missing_shards: List[int], shards_to_download: List[int], min_id: int,
+                                    max_id: int) -> None:
+        """Download and load the given missing shards.
 
-        Returns:
-            int: Dataset length.
+        Args:
+            missing_shards (List[int]): The missing shards.
+            shards_to_download (List[int]): List of shards to download by this worker.
+            min_id (int): Lowest sample ID of this partition.
+            max_id (int): Highest sample ID of this partition.
         """
-        return math.ceil(self.index.total_samples / dist.get_world_size())
+        for shard in missing_shards:
+            try:
+                self._download_shard(shard, shards_to_download)
+                self._load_shards([shard], min_id, max_id)
+            except Exception as e:
+                with self._lock:
+                    self._download_status = _DownloadStatus.FAILED
+                    self._download_exception = e
+                return
+        with self._lock:
+            self._download_status = _DownloadStatus.DONE
 
-    def _unpack_sample(self, data: bytes) -> Dict[str, Any]:
-        """Unpack a sample dict from raw bytes.
+    def download(self, blocking: bool = True, num_processes: Optional[int] = None) -> bool:
+        """Download and load all shards (optionally blocking, returns whether done).
 
-        First unpacks the str to raw bytes dict, then unpacks each field's raw bytes.
+        Bails out if has already been called.
 
         Args:
-            data (bytes): The packed bytes of the sample.
+            blocking (bool, default True): If blocking, downloads/loads all shards before returning. If non-blocking,
```

Comment:
Was this hack required because FFCV was calling `__getitem__` before the dataset was downloaded?

Stepping back, thoughts about making `_download` private and requiring that the caller iterate through the dataset to "cache" it rather than expose this functionality through the public api of the streaming dataset? Mainly thinking in terms of UX of why the caller should have to worry about downloading -- then the caller would need to be aware that the dataset is a streaming dataset. Besides this FFCV example, what are the use cases for having `download` be public?

Reply:
I think download will be useful in a number of cases. FFCV (it needs the full dataset on disk) and if someone wants to do some performance testing locally, making sure the dataset is available locally will be nice.